### PR TITLE
Broadcast `[0, contig)` range on replicator open and add want messages

### PR DESCRIPTION
**Pull request by kasperisager**

### CODE REVIEW

**Comment by kasperisager** [comment]
@mafintosh This is ready for review.

File: lib/replicator.js [hunk 1]
```diff
@@ -355,15 +355,7 @@ class Peer {
...
     this.sendSync()
-
-    const p = pages(this.core)
-
-    for (let index = 0; index < p.length; index++) {
-      this.wireBitfield.send({
-        start: index * this.core.bitfield.pageSize,
-        bitfield: p[index]
-      })
-    }
```
Comment (mafintosh) [inline lib/replicator.js:366]: ah we actually still need to send the bitfields, but only the ones with a start > contigPointer

**Comment by kasperisager** [comment]
@mafintosh This is ready for another review.

File: lib/replicator.js [hunk 2]
```diff
@@ -1299,6 +1343,28 @@ module.exports = class Replicator {
...
+      let end = i
+
+      while (end < length && this.core.bitfield.get(start + end) === true) end++
+
+      peer.wireRange.send({
+        start: start + i,
+        length: end - i
+      })
+    }
+
```
[the remark this answers is not in the document]
Comment (mafintosh) [reply lib/replicator.js:1364]: ie bitfield pages < start until >= end

File: test/replicate.js [hunk 3]
```diff
@@ -781,3 +781,57 @@ test('non-sparse snapshot during partial replication', async function (t) {
...
+
+  await t.test('discrete range', async function (t) {
+    s = replicate(b, c)
+    t.teardown(() => unreplicate(s))
+
+    await c.download({ blocks: [4, 6] }).downloaded()
+    t.pass('resolved')
+  })
+
+  await t.test('seek', async function (t) {
```
Comment (mafintosh) [inline test/replicate.js:831]: i think this fails if it happens as the first test here cuase the previous impact it

File: lib/replicator.js [hunk 4]
```diff
@@ -355,15 +366,7 @@ class Peer {
...
-
-    const p = pages(this.core)
-
-    for (let index = 0; index < p.length; index++) {
-      this.wireBitfield.send({
-        start: index * this.core.bitfield.pageSize,
-        bitfield: p[index]
-      })
-    }
+    this.broadcastRange(0, this.core.header.contiguousLength, false)
```
Comment (mafintosh) [inline lib/replicator.js:369]: we should prob only do this if contig > 0

File: test/replicate.js (hunk 3 above)
Comment (kasperisager) [reply test/replicate.js:831]: Oh shoot, I should have created separate copies of the third core 🤦

File: index.js [hunk 5]
```diff
@@ -534,6 +534,20 @@ module.exports = class Hypercore extends EventEmitter {
...
+
+    // When the contiguous length catches up, broadcast the non-sparse length
+    // to peers that haven't yet received it.
+    if (this.core !== null) {
+      const contig = this.core.header.contiguousLength
+
+      if (contig === this.core.tree.length) {
+        for (const peer of this.peers) {
+          if (peer.broadcastedNonSparse) continue
+
```
Comment (mafintosh) [inline index.js:546]: Shouldn’t you set the flag here also?

Comment (kasperisager) [reply index.js:546]: Yes 🤦

**Comment by mafintosh** [comment]
@kasperisager can you resolve the conflicts? then i'll merge tmw :)

**Comment by mafintosh** [comment]
excellent

**Pull request merged**